### ISSUE TITLE
Feature request - multiple condition groups (combined "ANY" and "ALL" search)

### ISSUE THREAD

**Issue by mlysle** [RealRaven2000/FiltaQuilla#113]
When setting up a filter, you can choose match "all of the following" (and) or "any of the following" (or), which is somewhat inflexible. In my case, I want to filter out certain emails. They always come from the same email address so I always want to match that, but the subject lines vary. But there's no way to express this in a single filter without getting creative with regex. And I don't want to filter out all emails from this sender in this case.

So what I am asking for is the ability to create some kind of condition groups ANDed together. Could be a premium feature. Hope that's understandable.

**Comment by RealRaven2000** [comment]
> When setting up a filter, you can choose match "all of the following" (and) or "any of the following" (or), which is somewhat inflexible. In my case, I want to filter out certain emails. They always come from the same email address so I always want to match that, but the subject lines vary. But there's no way to express this in a single filter without getting creative with regex. And I don't want to filter out all emails from this sender in this case.

You need to ask the core group to support this. The difficulty is in modifying the Filter editor in a meaningful way. Here is the mozilla bug that deals with this issue:

https://bugzilla.mozilla.org/show_bug.cgi?id=297852

One of my suggestions was to make filters "nestable" that is make one filter a condition for another, this way they could combine the different types without having to build a completely new UI

**Comment by RealRaven2000** [comment]
As a workaround, you could split the filter into 2 - use tags to create a result for the first condition query (it A and B and C tag email ABC), then query that in the 2nd filter (if tag=ABC or D ...). This is basically my low tech way of creating "nested" filters.

**Comment by jpmvferreira** [comment]
I would also like to support this feature. 

If I may in my head it would make sense to have logical operations in filters, which wouldn't require a new UI, only a new column which would be the "filter tag".

Let me exemplify this. In my case I have a filter to act only on my Inbox, that would filter two different emails to another folder. My idea was to do something like (where the first column is the filter tag):
```
Filter conditions:
A - Folder name - is - Inbox
B - To or CC - is - example@example.org
C - To or CC - is - example2@example.org

Perform these actions:
Move Message to - examplefolder
```

And then you would add new a line which would perform the logical operation:
```
Filter trigger:
(C ∨ B) ∧ A
```

So if either C or B is true, and A is true, then the whole expression is true, and the filter is activated.

Right now I'm not sure how I can solve this issue without creating a different filter for each e-mail, which is far from ideal, because adding a tag will raise the same issue as the folder.

**Comment by RealRaven2000** [comment]
> Let me exemplify this. In my case I have a filter to act only on my Inbox, that would filter two different emails to another folder. My idea was to do something like (where the first column is the filter tag):

If you want to avoid filters running outside of your inbox, there may be a simple workaround. I assume you are running filters on other folders manually, and you want to avoid this by restricing the running to the inbox only. Under "Apply Filter When" you could just uncheck [ ] manually run and enable [ ] getting new mail.

![image](https://user-images.githubusercontent.com/6314674/131308512-5d2cdd13-bd39-4dd3-a94f-47b0153b92c6.png)

> And then you would add new a line which would perform the logical operation:

that's not a practical solution, because of the way the conditions work - they really do not know of each other so you cannot create a new condition type that "looks at previous conditions". One would have to rewrite the whole filtering mechanism which is not something an Add-on can do - I think it runs in C++ for performance reasons. The only way this could work would be by the suggestion I made in [bug 297852](https://bugzilla.mozilla.org/show_bug.cgi?id=297852) - to have a filter condition that uses the result of another filter. 

But internally all the nuts + bolts are there - every search condition has these 3 boolean properties:
`booleanAnd`  = basically ∧ vs ∨
`beginsGrouping` = (
`endsGrouping` = )

the hard thing - that Thunderbird didn't manage in 16 years - is building a meaningful, easy to use, new UI to access these without making mistakes. (e.g. setting an open parenthesis and forgetting to close it).

**Comment by RealRaven2000** [comment]
@mlysle There is currently development work done on this area in the Thunderbird Clone Betterbird:

https://github.com/Betterbird/thunderbird-patches/issues/35

it is very likely that they might ship this with their next version BB102. It might also end upstream (in Thunderbird), but also likely that it's not going to be shipped this year there but only next year! We should definitely install Betterbird and see if might actually be superior in this and other areas; it is the most up-to-date fork for Thunderbird that I know of!

Here is a little preview (the blue operators were added by me as a suggestion for clarity)

![image](https://user-images.githubusercontent.com/6314674/163714860-8c6d607c-4f8c-46d3-8b48-21acd0c4aa85.png)

**Comment by RealRaven2000** [comment]
I am already testing a [Betterbird](https://www.betterbird.eu/) beta from @Betterbird :

https://bugzilla.mozilla.org/show_bug.cgi?id=297852#c67

![attachment](https://user-images.githubusercontent.com/6314674/164219252-4adecdb1-e92a-4dbd-809f-8d353d05078e.png)

**Comment by ThiloteE** [comment]
I will wait for the release, since I don't have the time to compile from source right now. I do have a question though:

Will this work with the filters:

> You can install the same dot release of Thunderbird in parallel and switch between Betterbird and Thunderbird on the same profile without problems and without needing the -allow-downgrade command line switch

Will my old filters work with Betterbird out of the box? If I create new filters, will I be able to use them with Thunderbird?

**Comment by ThiloteE** [comment]
(Sorry, for my last post, I posted it at the Betterbird repository instead! No need to answer this here)

**Comment by RealRaven2000** [comment]
> (Sorry, for my last post, I posted it at the Betterbird repository instead! No need to answer this here)

don't worry about it. I think an uplift of this patch to Thunderbird is possible but may take a while due to them being busy working on the big merge for ESR 102. Jörg has to fix stuff for saving the filters correctly and also even that wasn't the case opening the filters with an unpatched Thunderbird Filter Editor would definitely fail.  I will do some more testing this week to see whether you can Backup and restore the new filter formats with quickFilters without modifications, I will definitely implement it on my end.

**Comment by Betterbird** [comment]
Huh, it's all working in the latest build, https://www.betterbird.eu/downloads/get.php?os=win&lang=en-US&version=latest
Of course you need to keep https://github.com/Betterbird/thunderbird-patches/issues/35#issuecomment-1104040730 in mind.

**Comment by Betterbird** [comment]
New build with more compatibility now available, see https://github.com/Betterbird/thunderbird-patches/issues/35#issuecomment-1105503663.

**Comment by RealRaven2000** [comment]
Here is the current UI of Filter Editor from Betterbird (trying to host the screenshot here so I can share on Bugzilla):

![image](https://github.com/RealRaven2000/FiltaQuilla/assets/6314674/a7921a48-8fbf-4b54-ad7a-cd032915ea95)

**Comment by ThiloteE** [comment]
The screenshot is not visible at https://bugzilla.mozilla.org/show_bug.cgi?id=297852#c84

**Comment by RealRaven2000** [comment]
> The screenshot is not visible at https://bugzilla.mozilla.org/show_bug.cgi?id=297852#c84

yeah I did try imgur as well, but apparently the [Markdown documentation they link to](https://guides.github.com/features/mastering-markdown/) isn't 100% correct for Bugzilla. I remember when there was an interface for adding attachments....

**Comment by Betterbird** [comment]
Any reason why this group only has one member:
![image](https://github.com/RealRaven2000/FiltaQuilla/assets/86832723/a5627de2-3722-4404-a50f-7476c21dcd1a)

Therefore you don't need it.

**Comment by RealRaven2000** [comment]
> Any reason why this group only has one member: ![image](https://private-user-images.githubusercontent.com/86832723/331251689-a5627de2-3722-4404-a50f-7476c21dcd1a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NzA5NTUsIm5iZiI6MTcxNTg3MDY1NSwicGF0aCI6Ii84NjgzMjcyMy8zMzEyNTE2ODktYTU2MjdkZTItMzcyMi00NDA0LWE1MGYtNzQ3NmMyMWRjZDFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE2VDE0NDQxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlZTA5ZDI0MTc1MmExZjZjNTkzNDM2NDZmMGJhYzA2NzY4Y2UxOTEwMjkwYmZmMGVlMDUxZGRmZWE5ZjU0MzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.7AHMNU0Sb15qiaY0RiTWjWDz4XXPtnzT3ZVRfJiP6z8)
> 
> Therefore you don't need it.

yeah I know, was just a stupid example to show off the graphics... Also probably should have been a "Match any" in that group...

**Comment by Betterbird** [comment]
We have a picture here: https://www.betterbird.eu/media/img/complex-search-terms.png which shows the thing in action.
